### DISCUSSION TITLE
[Issue]: Tv Multiple Versions

### DISCUSSION THREAD

**Discussion by SenorSmartyPants** [jellyfin/jellyfin#7900] [General]
### Please describe your bug

I am the source of the reported issue #7855, This occurred because of my PR #7178. When developing it I noticed that it also added support for multiple versions of episodes. I liked that because I have some series with these, (aired and unaired pilots, 2 versions of Doctor Who The Five Doctors special) so I left it in. I should have documented the fact that this functionality was added and how to implement it on the user side with file/directory naming.

I would like JF to retain the ability to have multiple episode versions. I understand that this could cause some confusion based on the new behavior. I will update the documentation to include multiple tv versions, if we can keep this feature. I could also implement an option to turn this on if deemed necessary.

@cvium and @Bond-009 reviewed my tv extras pull. @crobibero turned if off in #7892

### Jellyfin Version

10.8.0

### Environment

```markdown
- OS:
- Virtualization:
- Clients:
- Browser:
- FFmpeg Version:
- Playback Method:
- Hardware Acceleration:
- Plugins:
- Reverse Proxy:
- Base URL:
- Networking:
- Storage:
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cvium** [comment]
We can't keep it without making significant changes and that won't be for 10.8. Feel free to discuss naming schemes and possible solutions/code changes. I will convert this to a discussion.

The movie naming scheme for multi-versions does not work for episodes.

**Comment by SenorSmartyPants** [comment]
What are the significant changes you think are necessary?

I will post naming examples in a couple hours. The movie style naming worked for my episode tests. I will provide examples in a bit.

**Reply by cvium** [reply]
It currently groups these three files if they aren't in a season folder

- My Show/My Show - S01E01 - 1080p.mkv
- My Show/My Show - S01E02 - 1080p.mkv
- My Show/My Show - S01E03 - 1080p.mkv

Clearly it needs to be smarter and that is not something we'll be doing in a 10.8.x release because it's liable to break a lot of people's setups.

**Reply by SenorSmartyPants** [reply]
Yes, That does seem to be a common problem/naming convention. I had all my tv shows with season folders or as `My Show S01E01 - 1080p.mkv` so I did not run into this while testing.

**Comment by SenorSmartyPants** [comment]
So here are the criteria to get multiple episodes working in 10.8.0. This relies on the unmodified movie version logic currently.

Episodes with multiple versions must be placed in a directory just for that episode. Directory name must exactly match what precedes the final space-dash-space to differentiate the versions in the file names.

```
- My Show/My Show S01E01/My Show S01E01 - Pilot.mkv
- My Show/My Show S01E01/My Show S01E01 - Pilot (extended).mkv
- My Show/My Show S01E02 - 1080p.mkv
- My Show/My Show S01E03 - 1080p.mkv
- My Show/Season 2/My Show S02E01/My Show S02E01 - Good.mkv
- My Show/Season 2/My Show S02E01/My Show S02E01 - Bad.mkv
```

Unfortunately as shown in the previous comments this can result in false positives when there is no season folder and space-dash-space is used after the series name. Then all episodes will be treated as multiple versions. 

Possible solutions to this include:
- Update filename parsing logic to be smarter about multiple versions. This would be the most development work, but could allow more naming conventions to trigger multiple versions. Probably not deployable until post 10.8 release.
- Create a library level option for tv libraries to enable multiple version support. Include link to documentation on proper filenaming to support this feature. Default would be not to enable to maintain pre-10.8.0 behavior. Much smaller development work and is opt-in for users. Minimal code base changes, could possibly be include in 10.8 point release, depending on opinion of JF dev community.
- User side: update tv naming convention to either always include season folder or exclude space-dash-space after series name. This is not desirable as breaks behavior and relies on users to fix it.

**Reply by ThomasPGH** [reply]
> Create a library level option for tv libraries to enable multiple version support. Include link to documentation on proper filenaming to support this feature. Default would be not to enable to maintain pre-10.8.0 behavior. Much smaller development work and is opt-in for users. Minimal code base changes, could possibly be include in 10.8 point release, depending on opinion of JF dev community.

This would be fantastic.

**Comment by SenorSmartyPants** [comment]
Since this issue wasn't caught by unit tests, one should be added. Maybe a list of episodes in a series directory and then verify the correct episode count.

**Comment by SenorSmartyPants** [comment]
I created a PR #8004 to update the parsing logic. There is no library level option to enable, it's always on like movie versions. I eliminated the false positives when episodes were in the series directory.

**Comment by almightiest** [comment]
I like these being in a separate folder, so it may not conflict with: https://features.jellyfin.org/posts/198/handle-movies-with-multiple-split-files-as-one-item

But there may be opportunity to tackle both here eventually?  Maybe detect if it's "Part 1" "Part 2" (or some other naming convention for multi parts), and then be sure if you choose Part 1 to play, that Part 2 would auto play in sequential order?  Worth it to have any tests around this type of naming?

Some examples from my library (Which works in Plex/Kodi but not Jellyfin):

```
Armchair.Thriller/Armchair.Thriller.S02/s02e29.Part1-armchair.thriller.chelsea.murders.feature-cd1.avi
Armchair.Thriller/Armchair.Thriller.S02/s02e29.Part2-armchair.thriller.chelsea.murders.feature-cd2.avi

The.Hollow.Crown/The.Hollow.Crown.S01/the.hollow.crown.s01e01.part1.dvdrip.xvid.cd1.avi
The.Hollow.Crown/The.Hollow.Crown.S01/the.hollow.crown.s01e01.part2.dvdrip.xvid.cd2.avi
```

**Reply by SenorSmartyPants** [reply]
What you are talking about is file stacking, which is not well documented in Jellyfin docs. But look at the file stacking section here [in emby docs](https://support.emby.media/support/solutions/articles/44001159102-movie-naming) and Jellyfin should work the same way.

This will most likely work for episodes when/if multiple episode versions is merged. But I have not tested that specifically. They might need to be in a directory just for that episode. Like I said, I haven't tested this.

**Reply by almightiest** [reply]
File stacking doesn't currently work for TV Episodes in Jellyfin, which is why that feature was requested.  I just know there could be some overlap depending on the solution you provide which is why I wanted you to see the other request.

**Comment by SenorSmartyPants** [comment]
Looks like [Emby](https://support.emby.media/support/solutions/articles/44001159110-tv-naming) now supports multiple versions for episodes.

**Reply by SenorSmartyPants** [reply]
Emby doesn't require a directory for each episode when doing multiple versions. I'm making this note to look into supporting that as well.

**Reply by GodTamIt** [reply]
Any updates to multiple versions without a directory?

I thought I might jump into some of the code this weekend to see if there's a reasonable way to refactor the parser to allow for sibling level versions without a directory (similar to what Plex does).

**Reply by SenorSmartyPants** [reply]
I have a PR that should support this. https://github.com/jellyfin/jellyfin/pull/8004

It hasn't been merged yet. And one user tried it recently and reported an issue, but I haven't checked that out just yet.

**Comment by VinnyStalck** [comment]
An alternative way that I see to fix this, is to add the feature to [support MKV files with multiple video streams](https://features.jellyfin.org/posts/942/support-multi-track-video-quality)
